### PR TITLE
Layout components for MDX content

### LINKED ISSUE #11044: Content authoring abstraction
## Why 
At the moment authoring design content is difficult and error prone.
We want to land at an abstraction that makes this process **fast** and **easy** for content authors. 

@tjonx and @zecarlostorre have provided strong and good design boundaries for how content can be [laid out](https://www.figma.com/file/vO0wToyBpYjeIwr8doAAWT/Uplift-polaris.shopify.com?type=design&node-id=1804-32995&mode=design&t=TIOTeGhOE6DsGZEk-4) 

Our goal is to implement an abstraction that allows content authors to lean into these strong design boundaries while: 
a) Not having to worry about layout implementation and browser behaviour 
b) Not having to learn and memorise too much magic incantations 
c) Not have to unlearn learned behaviour from the underlying technology (Markdown) 

## What 
The present solution that we've arrived at is an abstraction on top of the browsers [grid](https://developer.mozilla.org/en-US/docs/Web/CSS/grid) spec, specifically `[grid-template-areas](https://developer.mozilla.org/en-US/docs/Web/CSS/grid-template-areas)` and their ability to allow users to declare semantic areas where elements can be laid out. 

```
<LayoutSection card>
  <Text>
  </Text>

  <Small>
  </Small>

  <Medium>
  </Medium>
  
  <Large>
  </Large>
  
  <ExtraLarge>
  </ExtraLarge>
</LayoutSection>

<LayoutSection>
  <Text>
  </Text>

  <Small variant="do">
  </Small>

  <Small variant="dont">
  </Small>
</LayoutSection>
```

============

### PR DESCRIPTION
**Pull request by gwyneplaine**
### WHY are these changes introduced?

Fixes #11044 

### WHAT is this pull request doing?

- `Docs` Add new layout blocks for content authoring
- `Docs` refactor using-color to use the new layout primitives as a POC

```
<LayoutSection card>
  <Text>
    Content
  </Text>
  <Large>
    ![The admin background color](/images/design/colors/color-using-color-background@2x.png)
  </Large>
</LayoutSection>

<LayoutSection>
  <Medium variant="do">
    ![Background colors of the admin and of the sidenav](/images/design/colors/color-using-color-background-do@2x.png)

    Background color is always used in every admin interface.
  </Medium>

  <Medium variant="dont">
    ![Background color applied to a card](/images/design/colors/color-using-color-background-dont@2x.png)

    Use background color on surfaces or on individual elements.
  </Medium>
</LayoutSection>
```

### CODE REVIEW

**Comment by gwyneplaine** [comment]
This PR includes changes to `Grid.Cell` in `@shopify/polaris-react`, our solution here relies on being able to configure `grid-column-start` to conditionally horizontally offset our mdx layout primitives. 

For this solution to be mergeable into `main` we need to either: 
* Agree on and ship this change to the Grid component in @shopify/polaris (minor patch) 
* Find a solution that is comparable and doesn't involve this additional functionality. 

Personally I'm a proponent of our layout primitives letting us use the browser API with minimal / no guardrails, but I recognise that there's context here (especially around a11y) that I may not have. cc @kyledurand  @jesstelford

**Comment by kyledurand** [comment]
There's an ongoing [Grid RFC](https://github.com/Shopify/ui-api-design/discussions/185) that would be a good place to share these opinions. Since grid has been marked as stable we don't want to make any rash changes to it without knowing the APIs of other Grid components

**Comment by tjonx** [comment]
Do you think it would be wiser to just create specific components for the Polaris website instead, for this particular case? We'd move faster that way and won't be dependent on updating Polaris itself to get what we want.

**Comment by gwyneplaine** [comment]
> Do you think it would be wiser to just create specific components for the Polaris website instead, for this particular case? We'd move faster that way and won't be dependent on updating Polaris itself to get what we want.

@tjonx  we definitely can, just wanted to use this opportunity to call out potential changes in Grid. As that now looks like a slower process, I'm happy to duplicate Grid in docs for now with this additional change

**Comment by sam-b-rose** [comment]
> For this solution to be mergeable into main we need to either:
> * Agree on and ship this change to the Grid component in @shopify/polaris (minor patch)
> * **Find a solution that is comparable and doesn't involve this additional functionality.** 👈 

I like the MDX authoring solution here, though I think we should drop the `offset` prop and not add to the `Grid.Cell`. To "apply" an offset, I think we can use the strategy of inserting a blank/empty `<Small />` until we have more discussion around adding `columnStart` support to `Gird.Cell`.

File: polaris-react/src/components/Grid/components/Cell/Cell.tsx [hunk 1]
```diff
@@ -31,13 +31,15 @@ export interface CellProps {
    * https://polaris.shopify.com/components/layout-and-structure
    */
   area?: string;
+  colStart?: Cell;
```
Comment (sam-b-rose) [inline polaris-react/src/components/Grid/components/Cell/Cell.tsx]: What do you think about removing or spliting this addition off to a separate PR, then review the other MDX component changes (`Small`, `Medium`, etc)?

**Comment by yurm04** [comment]
are there any tradeoffs or downsides to making this change to the `polaris-react` component? If not, I don't have any strong opinions on merging in that change. 

If there are, I think we should establish a rule/principle about what it means to consume our polaris layout components in the website. Dogfooding our components in the website is great for working out bugs and finding the boundaries that we want to establish or expand on. But before we start making changes to components to fit the needs of the website, we should decide whether or not we want to expand the scope of our layout components beyond the Shopify admin.

Personally, I think our layout components should be as flexible as possible and we should continue adding support for "the 80%" whenever a situation calls for it. If this change gets us closer to 80% without compromising our ability to serve the shopify admin, I say go for it. But let's establish that as a principle of the layout components first so it can be used for future decision making and related discussions

**Comment by sam-b-rose** [comment]
@tjonx @kyledurand I've updated the PR to remove changes to `Grid.Cell` and updated the `using-color.mdx` which should have the expected authoring experience.

My current take-aways/thoughts:

- It makes sense, but only if you know what `Small`, `Medium`, and `Large` correspond to (1/3, 1/2, and 2/3 of the grid)
- We still lack some built-in card spacing here. The top-level Stack is still required, which isn't ideal imo
- The formatting is still very janky. It adds newlines in odd places and doesn't feel elegant

It is better than the very verbose version, though I'm not sure it is a strong improvement over the `Row` and `Column` method we prototyped before. Curious to hear your opinions, though!

**Pull request merged**